geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala

```diff
@@ -0,0 +1,135 @@
+      // The full-table scan block should be used to avoid the later case.
+      // Here a none is returned indicating that this guard does not stop the query
+      case None => None
+      case Some(intervals) =>
+        if (!intervals.forall(_.isBoundedBothSides) || intervals.isEmpty) {
+          Some(new IllegalArgumentException("At least one part of the query is temporally unbounded.  Add bounded temporal constraints."))
+        } else {
+          val queryDuration = duration(intervals.values)
+          guardLimits.find( _.sizeLimit >= spatialExtent).flatMap {
+            limit =>
+              if (queryDuration > limit.durationLimit) {
+                Some(new IllegalArgumentException(
+                  s"Query of this size exceeds maximum allowed filter duration of ${limit.durationLimit}: " +
+                    s"${filterToString(strategy.filter.filter)}. " +
+                    s"Try a small spatial extent or a shorter time range."))
+              } else {
+                None
+              }
+          }
+        }
+    }
+  }
+
+  override def close(): Unit = { }
+}
+object GraduatedQueryGuard extends LazyLogging {
+  val ConfigPath = "geomesa.guard.graduated"
+
+  /**
+   *
+   * @param sizeLimit in square degrees
+   * @param durationLimit Maximum duration for a query at or below the spatial size
+   */
+  case class SizeAndDuration(sizeLimit: Int, durationLimit: Duration)
+
+  def loadConfiguration(sft: SimpleFeatureType): Seq[SizeAndDuration]  = {
+    val conf = ConfigFactory.load()
+    try {
+      val guardConfig = conf.getConfigList(s"${ConfigPath}.${sft.getTypeName}")
+      buildLimits(guardConfig)
+    } catch {
+      case e: ConfigException =>
+        logger.error(s"Configuration for GraduatedQueryGuard not available or correct for type ${sft.getTypeName}.  " +
+          s"This query guard is disabled.")
+        Seq()
+    }
+  }
+
+  /**
+   * This function checks conditions on the limits.
+   * 1. Sizes must not be repeated.
+   * 2. Smaller sizes must allow for longer durations.
+   * @param limits Sequence of limits
+   * @return Sorted list of limits or throws an IllegalArgumentException
+   */
+  private def evaluateLimits(limits: Seq[SizeAndDuration]): Seq[SizeAndDuration] = {
+    val candidate = limits.sortBy(_.sizeLimit)
+    if (candidate.size > 1) {
+      candidate.sliding(2).foreach { case Seq(first, second) =>
+        if (first.sizeLimit == second.sizeLimit) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has repeated size: ${first.sizeLimit}")
+        } else if (first.durationLimit.compareTo(second.durationLimit) <= 0) {
+          throw new IllegalArgumentException(s"Graduated query guard configuration " +
+            s"has durations out of order: ${first.durationLimit} is less than ${second.durationLimit}")
+        }
+      }
+    }
+    if (candidate.last.sizeLimit != Int.MaxValue) {
```

Comment:
should probably make this .lastOption.exists or add a .isEmpty check for the degenerate case of an empty Seq

Reply:
Good call.  Adding an .isEmpty right ahead of this.